spec/integrations/rebalancing/message_order_with_sync_commit.rb

```diff
@@ -0,0 +1,162 @@
+          payload: { 'key' => SecureRandom.uuid }.to_json
+        }
+      end
+
+      Karafka::App.producer.produce_many_sync(messages)
+    rescue WaterDrop::Errors::ProducerClosedError
+      break
+    end
+
+    sleep(1)
+  end
+end
+
+other_consumer = Thread.new do
+  sleep(10)
+
+  consumer = setup_rdkafka_consumer
+  consumer.subscribe(DT.topic)
+
+  consumer.each do |message|
+    key = JSON.parse(message.payload)['key']
+
+    DT[:second_consumer] << [key, message.partition, message.offset]
+    DT[:consecutive_messages] << [key, message.partition, message.offset]
+    consumer.store_offset(message)
+
+    next unless DT[:contested_message].first&.payload&.dig('key') == key
+
+    # Stop at the contested message
+    break
+  end
+
+  consumer.commit(nil, false)
+  consumer.close
+
+  DT[:second_closed] = true
+end
+
+wait_until do
+  # Wait until the second consumer gets to the contested message and processes it
+  other_consumer.join && DT.data[:second_closed] == true && DT[:errors].count == 1
+end
+
+# Make sure that strict order is preserved (excluding the duplicated message)
+offsets_by_partition = DT[:consecutive_messages].group_by { |_, partition, _| partition }
+offsets_by_partition.each do |_, triples|
+  previous = nil
+
+  triples.each do |_, _, offset|
+    unless previous
+      previous = offset
+      next
+    end
+
+    # Contested message should be duplicated, since it should be processed by both consumers
+    contested_message_offset = DT[:contested_message].first.offset
+    next if previous == contested_message_offset && offset == contested_message_offset
+
+    assert_equal previous + 1, offset
+
+    previous = offset
+  end
+end
+
+# Second consumer should have started from the contested message, not from earlier ones
+messages_processed_by_second_consumer = DT[:second_consumer].select do |_, partition, _|
+  partition == DT[:contested_message].first.partition
+end
+second_consumer_offsets = messages_processed_by_second_consumer.map { |_, _, offset| offset }
+contested_message_key, = messages_processed_by_second_consumer.first
+
+assert_equal contested_message_key, DT[:contested_message].first.payload['key']
+
+assert_equal(true, second_consumer_offsets.none? { |offset| offset < DT[:contested_message].first.offset })
+
+# Contested message should have been reprocessed twice
+assert_equal(
+  DT[:consecutive_messages].select { |key, _, _| key == DT[:contested_message].first.payload['key'] }.count,
+  2
+)
+
+# #mark_as_consumed! will raise for the contested message instead of returning `true` ...
+absent_return_value = DT[:mark_as_consumed_return_values].find do |key, _|
+  key == DT[:contested_message].first.payload['key']
+end
+assert_equal absent_return_value, nil
+
+# ... it will raise Rdkafka::RdkafkaError instead
+assert_equal DT[:errors].count, 1
+consumption_error = DT[:errors].first
+assert_equal consumption_error.code, :unknown_member_id
+assert_equal consumption_error.class, Rdkafka::RdkafkaError
```

Comment:
Here's where the potential problem is: when the first consumer 'wakes up' and tries to synchronously commit the message after it has already lost the partition, #mark_as_consumed! raises Rdkafka::RdkafkaError with `unknown_member_id` code, which is not rescued by `Connection#internal_store_offset` and `false` is not returned. Is this intended?

Reply:
If `unknown_member_id` is not cought, it's definitely a bug. Let me try to create an isolated spec for it...

About your other things: yeah all over the place so I need to wrap by head around it :sweat_smile:

Reply:
@Aerdayne it's a bug. I was able to reduce the spec needed and will shortly commit the fix. Then we can get back to the original discussion.

Reply:
@Aerdayne fix: https://github.com/karafka/karafka/pull/1462